File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/WebAuthnRegistrationCallback.java
/*
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import android.annotation.TargetApi;
import android.content.Context;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;

import org.forgerock.android.auth.FRListener;
import org.forgerock.android.auth.InitProvider;
import org.forgerock.android.auth.Listener;
import org.forgerock.android.auth.Node;
import org.forgerock.android.auth.exception.WebAuthnResponseException;
import org.forgerock.android.auth.webauthn.WebAuthn;
import org.forgerock.android.auth.webauthn.WebAuthnListener;
import org.forgerock.android.auth.webauthn.WebAuthnRegistration;
import org.json.JSONException;
import org.json.JSONObject;

import lombok.Getter;
import lombok.NoArgsConstructor;

import static org.forgerock.android.auth.webauthn.WebAuthn.PUB_KEY_CRED_PARAMS;
import static org.forgerock.android.auth.webauthn.WebAuthn.WEBAUTHN_REGISTRATION;
import static org.forgerock.android.auth.webauthn.WebAuthn._ACTION;
import static org.forgerock.android.auth.webauthn.WebAuthn._PUB_KEY_CRED_PARAMS;
import static org.forgerock.android.auth.webauthn.WebAuthn._TYPE;

/**
 * A callback that allows some extra metadata to be sent in the response.
 * <p>
 * When serialized as JSON in an authenticate response, the {@link #getValue()} object will be the value of a single
 * {@literal data} output value, so for a value of {@code { "foo": "bar" }}, this would be output as:
 * <pre>
 * {@code
 * {
 *     "authId": "...",
 *     "callbacks": [
 *         // ...
 *         {
 *             "type": "MetadataCallback",
 *             "output": [
 *                 {
 *                     "name": "data",
 *                     "value": {
 *                         "_action": "webauthn_registration",
 *                         "challenge": "vlJwjd3nd76zjNfav2km/FKOgcIRa4BNHdVyasnPSb4=",
 *                         "attestationPreference": "none",
 *                         "userName": "e24f0d7c-a9d5-4a3f-a002-6f808210a8a3",
 *                         "userId": "ZTI0ZjBkN2MtYTlkNS00YTNmLWEwMDItNmY4MDgyMTBhOGEz",
 *                         "relyingPartyName": "ForgeRock",
 *                         "authenticatorSelection": "{\"userVerification\":\"preferred\",\"authenticatorAttachment\":\"platform\"}",
 *                         "_authenticatorSelection": {
 *                             "userVerification": "preferred",
 *                             "authenticatorAttachment": "platform"
 *                         },
 *                         "pubKeyCredParams": "[ { \"type\": \"public-key\", \"alg\": -7 }, { \"type\": \"public-key\", \"alg\": -257 } ]",
 *                         "_pubKeyCredParams": [
 *                             {
 *                                 "type": "public-key",
 *                                 "alg": -7
 *                             },
 *                             {
 *                                 "type": "public-key",
 *                                 "alg": -257
 *                             }
 *                         ],
 *                         "timeout": "60000",
 *                         "excludeCredentials": "",
 *                         "_excludeCredentials": [],
 *                         "displayName": "e24f0d7c-a9d5-4a3f-a002-6f808210a8a3",
 *                         "relyingPartyId": "id: \"openam.example.com\",",
 *                         "_relyingPartyId": "openam.example.com",
 *                         "_type": "WebAuthn"
 *                     }
 *                 }
 *             ]
 *         }
 *     ]
 * }
 * </pre>
 */
@NoArgsConstructor
@Getter
@TargetApi(24)
public class WebAuthnRegistrationCallback extends MetadataCallback implements WebAuthnCallback {

    @Keep
    public WebAuthnRegistrationCallback(JSONObject jsonObject, int index) {
        super(jsonObject, index);
    }

    /**
     * Check if this callback is {@link WebAuthnRegistrationCallback} Type
     *
     * @param value The callback raw data json.
     * @return True if this is a {@link WebAuthnRegistrationCallback} Type, else false
     */
    public static boolean instanceOf(JSONObject value) {
        //_action is provided AM version >= AM 7.1
        if (value.has(_ACTION)) {
            try {
                if (value.getString(_ACTION).equals(WEBAUTHN_REGISTRATION)) {
                    return true;
                }
            } catch (JSONException e) {
                //Should not happened
                return false;
            }
        }
        try {
            return (value.has(_TYPE) &&
                    value.getString(_TYPE).equals(WebAuthn.WEB_AUTHN) &&
                    (value.has(PUB_KEY_CRED_PARAMS) || value.has(_PUB_KEY_CRED_PARAMS)));
        } catch (JSONException e) {
            return false;
        }
    }

    @Override
    public String getType() {
        return "WebAuthnRegistrationCallback";
    }

    /**
     * Perform WebAuthn Registration.
     *
     * @param fragment The current {@link Fragment} that handle this callback
     * @param node     The Node returned from AM
     * @param listener Listener to listen for WebAuthn Registration Event
     */
    public void register(@NonNull Fragment fragment, Node node,
                         FRListener<Void> listener) {
        register(fragment.getContext(), fragment.getFragmentManager(), node, listener);
    }

    /**
     * Perform WebAuthn Registration with the current Activity, the Activity has to be of type
     * {@link FragmentActivity}
     *
     * @param node     The Node returned from AM
     * @param listener Listener to listen for WebAuthn Registration Event
     */
    public void register(Node node,
                         FRListener<Void> listener) {
        FragmentActivity fragmentActivity = InitProvider.getCurrentActivityAsFragmentActivity();
        register(fragmentActivity.getApplicationContext(), fragmentActivity.getSupportFragmentManager(), node, listener);
    }

    protected WebAuthnRegistration getWebAuthnRegistration() throws JSONException {
        return new WebAuthnRegistration(getValue());
    }

    private void register(@NonNull Context context, @NonNull FragmentManager fragmentManager,
                          @NonNull Node node, FRListener<Void> listener) {
        try {
            getWebAuthnRegistration().register(context,
                    fragmentManager,
                    new WebAuthnListener() {
                        @Override
                        public void onSuccess(String result) {
                            setHiddenCallbackValue(node, result);
                            Listener.onSuccess(listener, null);
                        }

                        @Override
                        public void onException(WebAuthnResponseException e) {
                            setHiddenCallbackValue(node, e.toServerError());
                            Listener.onException(listener, e);
                        }

                        @Override
                        public void onUnsupported(WebAuthnResponseException e) {
                            setHiddenCallbackValue(node, "unsupported");
                            Listener.onException(listener, e);
                        }

                        @Override
                        public void onException(Exception e) {
                            setHiddenCallbackValue(node, "ERROR::UnknownError:" + e.getMessage());
                            Listener.onException(listener, e);
                        }
                    });
        } catch (UnsupportedOperationException e) {
            setHiddenCallbackValue(node, "unsupported");
            Listener.onException(listener, e);
        } catch (Exception e) {
            Listener.onException(listener, e);
        }
    }
}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/WebAuthnRegistrationCallback.kt
/*
 * Copyright (c) 2022 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import android.annotation.TargetApi
import android.content.Context
import androidx.annotation.Keep
import com.google.android.gms.fido.fido2.api.common.ResidentKeyRequirement
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.forgerock.android.auth.FRListener
import org.forgerock.android.auth.Listener
import org.forgerock.android.auth.Node
import org.forgerock.android.auth.webauthn.WebAuthn
import org.forgerock.android.auth.webauthn.WebAuthnRegistration
import org.forgerock.android.auth.webauthn.cloneWith
import org.json.JSONException
import org.json.JSONObject

/**
 * A callback that handle WebAuthnRegistration Node.
 */
@TargetApi(24)
open class WebAuthnRegistrationCallback : MetadataCallback, WebAuthnCallback {

    @Keep
    @JvmOverloads
    constructor(jsonObject: JSONObject, index: Int) : super(jsonObject, index)

    @Keep
    @JvmOverloads
    constructor() : super()

    override fun getType(): String {
        return "WebAuthnRegistrationCallback"
    }

    private var residentKeyRequirement: String? = null

    open fun setResidentKeyRequirement(requirement: ResidentKeyRequirement) {
        residentKeyRequirement = requirement.name
    }


    /**
     * Perform WebAuthn Registration
     *
     * @param context The application context
     * @param node The current Node
     */
    suspend fun register(context: Context, deviceName: String? = null, node: Node) {
        try {
            val webAuthnRegistration = getWebAuthnRegistration()
            //Override the ResidentKeyRequirement
            residentKeyRequirement?.let {
                webAuthnRegistration.options = webAuthnRegistration.options.cloneWith(
                    ResidentKeyRequirement.valueOf(it))
            }
            var result = webAuthnRegistration.register(context)
            deviceName?.apply { result += "::$deviceName" }
            setHiddenCallbackValue(node, result);
        } catch (e: Exception) {
            setErrorRethrow(node, e)
        }
    }

    /**
     * Perform WebAuthn Registration
     *
     * @param node     The current Node
     * @param listener Listener to listen for WebAuthn Registration Event
     */
    @JvmOverloads
    fun register(context: Context, deviceName: String? = null, node: Node,
                 listener: FRListener<Void?>) {
        val scope = CoroutineScope(Dispatchers.Default)
        scope.launch {
            try {
                register(context, deviceName, node)
                Listener.onSuccess(listener, null)
            } catch (e: Exception) {
                Listener.onException(listener, e)
            }
        }
    }


    @Throws(Exception::class)
    open fun getWebAuthnRegistration() = WebAuthnRegistration(value)

    companion object {
        /**
         * Check if this callback is [WebAuthnRegistrationCallback] Type
         *
         * @param value The callback raw data json.
         * @return True if this is a [WebAuthnRegistrationCallback] Type, else false
         */
        @JvmStatic
        fun instanceOf(value: JSONObject): Boolean {
            //_action is provided AM version >= AM 7.1
            if (value.has(WebAuthn._ACTION)) {
                try {
                    if (value.getString(WebAuthn._ACTION) == WebAuthn.WEBAUTHN_REGISTRATION) {
                        return true
                    }
                } catch (e: JSONException) {
                    //Should not happened
                    return false
                }
            }
            return try {
                value.has(WebAuthn._TYPE) && value.getString(WebAuthn._TYPE) == WebAuthn.WEB_AUTHN && (value.has(
                    WebAuthn.PUB_KEY_CRED_PARAMS) || value.has(WebAuthn._PUB_KEY_CRED_PARAMS))
            } catch (e: JSONException) {
                false
            }
        }
    }
}